File: app/schemas/onboarding.py
from pydantic import BaseModel, EmailStr
from typing import Optional, Any
from datetime import datetime
from app.models.onboarding import SessionStatus, CustomerType, ImportSource


class SessionCreate(BaseModel):
    industry_id: str
    applicant_name: str
    applicant_email: EmailStr
    applicant_phone: Optional[str] = None
    applicant_company: Optional[str] = None
    customer_type: CustomerType = CustomerType.individual
    created_by: Optional[str] = None


class SessionSummary(BaseModel):
    id: int
    session_id: str
    industry_id: str
    applicant_name: str
    applicant_email: str
    applicant_phone: Optional[str]
    customer_type: str
    status: str
    current_step: int
    total_steps: int
    completion_pct: float
    documents_uploaded: int
    reminders_sent: int
    sanctions_match: Optional[bool]
    risk_level: Optional[str]
    risk_score: Optional[float]
    customer_id: Optional[str]
    kyc_id: Optional[str]
    source: str
    invite_sent_at: Optional[datetime]
    invite_opened_at: Optional[datetime]
    completed_at: Optional[datetime]
    created_at: Optional[datetime]

    class Config:
        from_attributes = True


class SessionDetail(SessionSummary):
    invite_token: str
    invite_expires_at: Optional[datetime]
    collected_data: Optional[dict]
    batch_id: Optional[str]
    created_by: Optional[str]

    class Config:
        from_attributes = True


class AuditLogEntry(BaseModel):
    id: int
    event_type: str
    event_data: Optional[dict]
    actor: Optional[str]
    ip_address: Optional[str]
    created_at: Optional[datetime]

    class Config:
        from_attributes = True


class PortalTokenResponse(BaseModel):
    session_id: str
    industry_id: str
    applicant_name: str
    applicant_email: str
    applicant_company: Optional[str]
    customer_type: str
    status: str
    current_step: int
    total_steps: int
    completion_pct: float
    collected_data: Optional[dict]
    invite_expires_at: Optional[datetime]
    steps: list[dict]


class StepSubmit(BaseModel):
    step: int
    data: dict[str, Any]


class BatchSummary(BaseModel):
    batch_id: str
    industry_id: str
    source: str
    file_name: Optional[str]
    total_rows: int
    success_rows: int
    error_rows: int
    errors: Optional[list]
    created_at: Optional[datetime]

    class Config:
        from_attributes = True


class PipelineStats(BaseModel):
    total: int
    invited: int
    opened: int
    in_progress: int
    completed: int
    rejected: int
    expired: int
    avg_completion_pct: float
    sanctions_matches: int

File: web/lib/api/onboarding.ts
// Typed onboarding-resource client — thirteenth pilot for the shared
// lib/api/client.ts pattern. Covers app/onboarding/page.tsx's and
// components/Onboarding/BulkUpload.tsx's /api/v1/onboarding/* call
// sites — the ones that go through apiFetch() (Bearer token/CSRF,
// authenticated operator dashboard).
//
// Deliberately excludes app/onboarding/[token]/page.tsx and
// components/Onboarding/CustomerPortal.tsx, which hit
// /api/v1/onboarding/portal/{token}/* via plain, unauthenticated
// fetch() — a public, token-based flow for the applicant's own browser
// session, not the operator's. Genuinely different auth model, same
// discipline as excluding onboarding's public pages from the
// authenticated resource client throughout this migration.
//
// Types mirror app/schemas/onboarding.py's SessionSummary/SessionDetail/
// SessionCreate/BatchSummary/PipelineStats.

import { apiGet, apiPost } from './client'

export interface OnboardingSession {
  id: number
  session_id: string
  industry_id: string
  applicant_name: string
  applicant_email: string
  applicant_phone?: string | null
  customer_type: string
  status: string
  current_step: number
  total_steps: number
  completion_pct: number
  documents_uploaded: number
  reminders_sent: number
  sanctions_match?: boolean | null
  risk_level?: string | null
  risk_score?: number | null
  customer_id?: string | null
  kyc_id?: string | null
  source: string
  invite_sent_at?: string | null
  invite_opened_at?: string | null
  completed_at?: string | null
  created_at?: string | null
}

export interface OnboardingSessionDetail extends OnboardingSession {
  invite_token: string
  invite_expires_at?: string | null
  collected_data?: Record<string, unknown> | null
  batch_id?: string | null
  created_by?: string | null
}

export interface CreateSessionInput {
  applicant_name: string
  applicant_email: string
  applicant_phone?: string
  applicant_company?: string
  customer_type?: string
  industry_id: string
}

export interface PipelineStats {
  total: number
  invited: number
  opened: number
  in_progress: number
  completed: number
  rejected: number
  expired: number
  avg_completion_pct: number
  sanctions_matches: number
}

export interface BatchImportError {
  row?: number | null
  data?: unknown
  error: string
}

export interface BatchSummary {
  batch_id: string
  industry_id: string
  source: string
  file_name?: string | null
  total_rows: number
  success_rows: number
  error_rows: number
  errors?: BatchImportError[] | null
  created_at?: string | null
}

/** GET /onboarding/sessions */
export function listOnboardingSessions(industryId: string, limit?: number): Promise<OnboardingSession[]> {
  const q = limit != null ? `&limit=${limit}` : ''
  return apiGet(`/api/v1/onboarding/sessions?industry_id=${encodeURIComponent(industryId)}${q}`)
}

/** GET /onboarding/stats */
export function getOnboardingStats(industryId: string): Promise<PipelineStats> {
  return apiGet(`/api/v1/onboarding/stats?industry_id=${encodeURIComponent(industryId)}`)
}

/** POST /onboarding/sessions/{id}/remind */
export function sendOnboardingReminder(sessionId: string): Promise<unknown> {
  return apiPost(`/api/v1/onboarding/sessions/${sessionId}/remind`)
}

/** POST /onboarding/sessions/{id}/cancel */
export function cancelOnboardingSession(sessionId: string): Promise<OnboardingSessionDetail> {
  return apiPost(`/api/v1/onboarding/sessions/${sessionId}/cancel`)
}

/** POST /onboarding/sessions */
export function createOnboardingSession(payload: CreateSessionInput): Promise<OnboardingSessionDetail> {
  return apiPost('/api/v1/onboarding/sessions', payload)
}

/** POST /onboarding/import/csv or /onboarding/import/excel, by file extension. */
export function importOnboardingFile(file: File, industryId: string): Promise<BatchSummary> {
  const ext = file.name.split('.').pop()?.toLowerCase()
  const endpoint = ext === 'xlsx' || ext === 'xls' ? '/import/excel' : '/import/csv'
  const form = new FormData()
  form.append('industry_id', industryId)
  form.append('file', file)
  return apiPost(`/api/v1/onboarding${endpoint}`, undefined, { body: form })
}
